lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/FileSystemInputPartition.cs

```diff
@@ -54,26 +56,73 @@ internal sealed class FileSystemInputPartition<T> : IInputPartition<T>, IDisposa
             _id = id;
             _fileSystem = fileSystem;
             _fileSerializer = fileSerializer;
-            _filePaths = filePaths;
             _tempFileCreator = tempFileCreator;
-            _isInitialized = false;
-            _copyToLocal = copyToLocal;
+            _filePaths = filePaths;
+
+            if (!copyToLocal)
+            {
+                // Implies that the files are already local.
+                _localFiles = Optional<ISet<string>>.Of(filePaths);
+            }
+            else
+            {
+                _localFiles = Optional<ISet<string>>.Empty();
+            }
         }
 
         public string Id
         {
             get { return _id; }
         }
 
-        private void Initialize()
+        public void Cache()
         {
             lock (_lock)
             {
-                if (!_isInitialized)
+                if (_localFiles.IsPresent())
                 {
-                    CopyFromRemote();
-                    _isInitialized = true;
+                    return;
                 }
+
```

Comment:
I was more concerned about the relative length of this method. The `if/else` would be the only thing in this method, and the actual download logic would be in the helper methods:

```
 if (!_localFiles.IsPresent()){ 
  Download();
}
```